velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.cpp

```diff
@@ -0,0 +1,370 @@
+  VELOX_CHECK_GE(indexBitLength, 1, "indexBitLength must be >= 1.");
+}
+
+uint32_t numberOfTrailingZeros(uint64_t hash, uint32_t indexBitLength) {
+  // Set the lowest bit in the prefix to ensure value is non-zero
+  constexpr uint32_t kBitWidth = sizeof(uint64_t) * 8;
+  uint64_t value = hash | (1ULL << (kBitWidth - indexBitLength));
+
+  // Safe because value is guaranteed non-zero
+  return static_cast<uint32_t>(__builtin_ctzll(value));
+}
+} // namespace
+
+using Allocator = facebook::velox::StlAllocator<int8_t>;
+
+SfmSketch::SfmSketch(HashStringAllocator* allocator)
+    : bits_(Allocator(allocator)) {}
+
+void SfmSketch::initialize(uint32_t buckets, uint32_t precision) {
+  auto indexBitLen = indexBitLength(buckets);
+  validateIndexBitLength(indexBitLen);
+
+  VELOX_CHECK_LE(
+      precision + indexBitLen,
+      64,
+      "Precision + indexBitLength cannot exceed 64");
+  indexBitLength_ = indexBitLen;
+  precision_ = precision;
+  bits_.resize(velox::bits::nbytes(buckets * precision));
+}
+
+void SfmSketch::addHash(uint64_t hash) {
+  auto bucketIndex = computeIndex(hash, indexBitLength_);
+  // Cap the number of trailing zeros to precision - 1, to avoid out of
+  // bound.
+  auto zeros =
+      std::min(precision_ - 1, numberOfTrailingZeros(hash, indexBitLength_));
+  setBitTrue(bucketIndex, zeros);
+}
+
+void SfmSketch::addIndexAndZeros(uint32_t bucketIndex, uint32_t zeros) {
+  auto numOfbuckets = numberOfBuckets(indexBitLength_);
+  VELOX_CHECK_GE(bucketIndex, 0, "Bucket index out of range.");
+  VELOX_CHECK_LT(bucketIndex, numOfbuckets, "Bucket index out of range.");
+  VELOX_CHECK_GE(zeros, 0, "Zeros must be GREATER than or equal to 0.");
+  VELOX_CHECK_LE(zeros, 64, "Zeros must be LESS than or equal to 64.");
+
+  // count of zeros in range [0, precision - 1]
+  zeros = std::min(precision_ - 1, zeros);
+  setBitTrue(bucketIndex, zeros);
+}
+
+uint32_t SfmSketch::computeIndex(uint64_t hash, uint32_t indexBitLength) {
+  validateIndexBitLength(indexBitLength);
+  constexpr uint32_t kBitWidth = sizeof(uint64_t) * 8;
+  return static_cast<uint32_t>(hash >> (kBitWidth - indexBitLength));
+}
+
+uint32_t SfmSketch::numberOfBuckets(uint32_t indexBitLength) {
+  validateIndexBitLength(indexBitLength);
+  return 1 << indexBitLength;
+}
+
+uint32_t SfmSketch::indexBitLength(uint32_t numberOfBuckets) {
+  VELOX_CHECK_GT(numberOfBuckets, 0, "numberOfBuckets must be greater than 0.");
+  VELOX_CHECK_EQ(
+      numberOfBuckets & (numberOfBuckets - 1),
+      0,
+      "numberOfBuckets must be power of 2.");
+  return static_cast<uint32_t>(std::log2(numberOfBuckets));
+}
+
+double SfmSketch::calculateRandomizedResponseProbability(double epsilon) {
+  if (epsilon == kNonPrivateEpsilon) {
+    return 0.0;
+  }
+  return 1.0 / (1.0 + exp(epsilon));
+}
+
+void SfmSketch::mergeWith(const SfmSketch& other) {
+  MersenneTwisterRandomizationStrategy randomizationStrategy;
+  mergeWith(other, &randomizationStrategy);
+}
+
+void SfmSketch::mergeWith(
```

Comment:
It looks like we always call this method with MersenneTwisterRandomizationStrategy. If so, this method is not needed.